lib/active_merchant/billing/gateways/shift4.rb

```diff
@@ -0,0 +1,241 @@
+module ActiveMerchant #:nodoc:
+  module Billing #:nodoc:
+    class Shift4Gateway < Gateway
+      self.test_url = 'https://utgapi.shift4test.com/api/rest/v1/'
+      self.live_url = 'https://utg.shift4api.net/api/rest/v1/'
+
+      self.supported_countries = ['US']
+      self.default_currency = 'USD'
+      self.supported_cardtypes = %i[visa master american_express discover]
+
+      self.homepage_url = 'https://utg.shift4api.net'
+      self.display_name = 'Shift4'
+
+      STANDARD_ERROR_CODE_MAPPING = {}
+
+      def initialize(options = {})
+        requires!(options, :client_guid, :auth_token)
+        super
+      end
+
+      def access_token(options = {})
+        post = {}
+        add_credentials(post, options)
+        commit('accesstoken', post, request_headers(options))
+      end
+
+      def purchase(money, options = {})
+        post = {}
+        add_datetime(post, options)
+        api_options(post, options)
+        add_invoice(post, money, options)
+        add_clerk(post, money, options)
+        add_transaction(post, options)
+        add_card(post, options)
+
+        commit('sale', post, request_headers(options))
+      end
+
+      def authorize(money, options = {})
+        post = {}
+        add_datetime(post, options)
+        add_invoice(post, money, options)
+        add_clerk(post, money, options)
+        add_transaction(post, options)
+        add_card(post, options)
+        add_customer(post, options)
+
+        commit('authorization', post, request_headers(options))
+      end
+
+      def capture(money, options = {})
+        post = {}
+        add_datetime(post, options)
+        add_invoice(post, money, options)
+        add_clerk(post, money, options)
+        add_transaction(post, options)
+        add_card(post, options)
+
+        commit('capture', post, request_headers(options))
+      end
+
+      def refund(money, options = {})
+        post = {}
+        add_datetime(post, options)
+        api_options(post, options)
+        add_invoice(post, money, options)
+        add_clerk(post, money, options)
+        add_transaction(post, options)
+        add_device(post, options)
+        add_p2pe(post, options)
+        add_emv(post, options)
+        add_customer(post, options)
+        add_card(post, options)
+
+        commit('refund', post, request_headers(options))
+      end
+
+      def void(options = {})
+        post = {}
+        commit('invoice', post, request_headers(options))
+      end
+
+      def verify(credit_card, options = {})
+        MultiResponse.run(:use_first_response) do |r|
+          r.process { authorize(100, credit_card, options) }
+          r.process(:ignore_result) { void(r.authorization, options) }
+        end
+      end
+
+      private
+
+      def add_clerk(post, money, options)
+        post[:clerk] = {}
+        post[:clerk][:numericId] = options[:clerk][:numeric_id]
```

Comment:
We should avoid nesting options when there's not a compelling reason to do so; this could be something like `options[:clerk_id]`

Reply:
@therufs  Done.